### PR TITLE
Kernel harmonization

### PR DESCRIPTION
**Pull request by antoine-levitt**

### CODE REVIEW

File: src/terms/hartree.jl [hunk 1]
```diff
@@ -19,7 +19,7 @@ Hartree(; scaling_factor=1) = Hartree(scaling_factor)
 
 struct TermHartree <: Term
     basis::PlaneWaveBasis
-    scaling_factor::Real
+    scaling_factor::Real # scaling factor, absorbed into poisson_green_coeffs
```
Comment (mfherbst) [inline src/terms/hartree.jl:22]: Space before `#`

File: src/terms/terms.jl [hunk 2]
```diff
@@ -67,5 +50,63 @@ breaks_symmetries(term_builder::Magnetic) = true
 breaks_symmetries(term_builder::ExternalFromReal) = true
 breaks_symmetries(term_builder::ExternalFromFourier) = true
 
-### Builders are objects X that store the term parameters, and produce a
-# XTerm <: Term when instantiated with a `basis`
+
+
+
```
Comment (mfherbst) [inline src/terms/terms.jl:55]: Too many empty lines.

File: src/terms/terms.jl [hunk 3]
```diff
@@ -67,5 +50,63 @@ breaks_symmetries(term_builder::Magnetic) = true
...
+            f += ft
+        end
+    end
+    f
+end
+forces(scfres) = forces(scfres.ham.basis, scfres.ψ, scfres.occupation, ρ=scfres.ρ)
+
+
+
+
```
Comment (mfherbst) [inline src/terms/terms.jl:80]: Too many empty lines.

File: src/terms/terms.jl [hunk 4]
```diff
@@ -67,5 +50,63 @@ breaks_symmetries(term_builder::Magnetic) = true
...
+"""
+    compute_kernel(term::Term; kwargs...)
+
+Computes a matrix representation of the full response kernel
+(derivative of potential with respect to density) in real space.
+"""
+function compute_kernel(term::Term; kwargs...)
+    false  # "strong zero": false + something = something
```
Comment (mfherbst) [inline src/terms/terms.jl:88]: What's the difference to using `0`

File: src/terms/terms.jl [hunk 5]
```diff
@@ -67,5 +50,63 @@ breaks_symmetries(term_builder::Magnetic) = true
...
+end
+"""
+    apply_kernel(term::Term, dρ; kwargs...)
+
+Computes the potential response to a perturbation dρ in real space
```
Comment (mfherbst) [inline src/terms/terms.jl:93]: Maybe mention that `dρ` is only a real-space array.

File: test/kernel.jl [hunk 6]
```diff
@@ -3,67 +3,33 @@ using Test
...
+        pot_minus = ene_ops(term, nothing, nothing; ρ=ρ_minus).ops[1].potential
+        ρ_plus = from_real(basis, ρ0.real .+ ε * dρ)
+        pot_plus = ene_ops(term, nothing, nothing; ρ=ρ_plus).ops[1].potential
+        dV = (pot_plus - pot_minus) / (2ε)
+
+        dV_apply = DFTK.apply_kernel(term, dρ; ρ=ρ0)
+        ker = DFTK.compute_kernel(term; ρ=ρ0)
+        dV_compute = reshape(ker * vec(dρ), size(dρ))
+        @test norm(dV - dV_apply) < 1e-6
+        @test norm(dV - dV_compute) < 1e-6
```
Comment (mfherbst) [inline test/kernel.jl:33]: Don't hard-code. Should depend on ε.

File: src/terms/terms.jl (hunk 4 above)
Comment (antoine-levitt) [reply src/terms/terms.jl:88]: Try `-0.0 + 0` vs `-0.0 + false`. false also allows more optimizations.

Comment (mfherbst) [reply src/terms/terms.jl:88]: Is this documented somewhere?

Comment (antoine-levitt) [reply src/terms/terms.jl:88]: It's part IEEE-whatever semantics, part julia promotion rules. Not sure of any explicit docs

Comment (mfherbst) [reply src/terms/terms.jl:88]: ok. Thanks ... good to know nevertheless.

**Pull request merged**